diff --git a/src/cli.js b/src/cli.js
index ca28b1ee1464eeb3f41af007b887a110ebee74e3..e4784a61683424a4b6b481c5bcec1ae9bb5eefa0 100644
--- a/src/cli.js
+++ b/src/cli.js
@@ -30,9 +30,10 @@ prog
 		'--tsconfig',
 		'A path to a tsconfig or jsconfig file. When not provided, searches for the next upper tsconfig/jsconfig in the workspace path.'
 	)
+	.option('-c, --cwd', 'The current working directory')
 	.action(async (args) => {
 		try {
-			const config = await load_config();
+			const config = await load_config({cwd: args.cwd});
 
 			// @ts-expect-error
 			if (config.package) {
diff --git a/src/index.js b/src/index.js
index fb540e7bae5125f481e6afc6aab8f8027d3d6303..d94f091f79299e9173c75a1a11e864f5db77b704 100644
--- a/src/index.js
+++ b/src/index.js
@@ -28,17 +28,36 @@ async function do_build(options, analyse_code) {
 		throw new Error(`${path.relative('.', input)} does not exist`);
 	}
 
-	rimraf(temp);
-	mkdirp(temp);
-
-	const files = scan(input, extensions);
+	const tmp_input = path.join(input, 'tmp_input')
+
+ 	rimraf(temp);
+	rimraf(tmp_input)
+ 	mkdirp(temp);
+
+	mkdirp(tmp_input)
+	copy(input, tmp_input, {
+		filter: (p) => 
+			!p.includes('node_modules') && 
+			!p.includes('dist') && 
+			!p.includes('tmp_input') && 
+			!p.includes('public') &&
+			!p.endsWith('.stories.svelte') && 
+			!p.endsWith('.test.ts') && 
+			!p.endsWith('.spec.ts') && 
+			!p.endsWith('.md') &&
+			!p.endsWith('package.json') &&
+			!p.endsWith('.gitignore') &&
+			!p.endsWith('.svelte-kit')
+	})
+
+	const files = scan(tmp_input, extensions);
 
 	if (options.types) {
-		await emit_dts(input, temp, output, options.cwd, alias, files, tsconfig);
+		await emit_dts(tmp_input, temp, output, options.cwd, alias, files, tsconfig);
 	}
 
 	for (const file of files) {
-		await process_file(input, temp, file, options.config.preprocess, alias, tsconfig, analyse_code);
+		await process_file(tmp_input, temp, file, options.config.preprocess, alias, tsconfig, analyse_code);
 	}
 
 	if (!options.preserve_output) {
@@ -47,11 +66,14 @@ async function do_build(options, analyse_code) {
 
 	mkdirp(output);
 	copy(temp, output);
+	rimraf(path.join(temp, ".."))
+	rimraf(tmp_input)
 
 	console.log(
 		colors
 			.bold()
-			.green(`${path.relative(options.cwd, input)} -> ${path.relative(options.cwd, output)}`)
+			.green(`${path.relative(path.join(options.cwd, "..", ".."), input)} -> ${path.relative(options.cwd, output)}`)
+
 	);
 }
 
diff --git a/src/typescript.js b/src/typescript.js
index 81c0f348e6358349c6ac11196f66ce34e9a766f2..3a39726d82df844bb0bfd3c5b11ca0843c06c61a 100644
--- a/src/typescript.js
+++ b/src/typescript.js
@@ -22,6 +22,7 @@ import { load_pkg_json } from './config.js';
  */
 export async function emit_dts(input, output, final_output, cwd, alias, files, tsconfig) {
 	const tmp = `${output}/__package_types_tmp__`;
+	const root = path.resolve(cwd, "..", "..")
 	rimraf(tmp);
 	mkdirp(tmp);
 
@@ -40,7 +41,7 @@ export async function emit_dts(input, output, final_output, cwd, alias, files, t
 		svelteShimsPath: no_svelte_3
 			? require.resolve('svelte2tsx/svelte-shims-v4.d.ts')
 			: require.resolve('svelte2tsx/svelte-shims.d.ts'),
-		declarationDir: path.relative(cwd, tmp),
+		declarationDir: path.relative(root, tmp),
 		tsconfig
 	});
 
